rest-api/src/main/java/gov/cms/qpp/conversion/api/controllers/SkeletalQrdaController.java

```diff
@@ -0,0 +1,127 @@
+import gov.cms.qpp.conversion.api.services.QrdaService;
+import gov.cms.qpp.conversion.api.services.ValidationService;
+
+/**
+ * Controller to handle uploading files for QRDA-III Conversion
+ */
+@RestController
+@CrossOrigin(allowCredentials = "true")
+public abstract class SkeletalQrdaController<T> {
+
+	private static final Logger API_LOG = LoggerFactory.getLogger(SkeletalQrdaController.class);
+	private static final int MAX_PURPOSE_LENGTH = 25;
+
+	protected final QrdaService qrdaService;
+	protected final ValidationService validationService;
+	protected final AuditService auditService;
+
+	/**
+	 * init dependencies
+	 *
+	 * @param qrdaService {@link QrdaService} to perform QRDA to QPP conversion
+	 * @param validationService {@link ValidationService} to perform post conversion validation
+	 * @param auditService {@link AuditService} to persist audit information
+	 */
+	public SkeletalQrdaController(QrdaService qrdaService, ValidationService validationService, AuditService auditService) {
+		this.qrdaService = qrdaService;
+		this.validationService = validationService;
+		this.auditService = auditService;
+	}
+
+	/**
+	 * Endpoint to transform an uploaded file into a valid or error json response
+	 *
+	 * @param file Uploaded file
+	 * @param purpose the purpose for the conversion
+	 * @return Valid json or error json content
+	 */
+	@PostMapping
+	public ResponseEntity<T> uploadQrdaFile(
+		@RequestParam(name = "file") MultipartFile file,
+		@RequestHeader(required = false, name = "Purpose") String purpose) {
+		String originalFilename = file.getOriginalFilename();
+
+		if (!StringUtils.isEmpty(purpose)) {
+			if (purpose.length() > MAX_PURPOSE_LENGTH) {
+				throw new InvalidPurposeException("Given Purpose (header) is too large. Max length is "
+						+ MAX_PURPOSE_LENGTH + ", yours was " + purpose.length());
+			}
+			API_LOG.info("Conversion request received for " + purpose);
+		} else {
+			purpose = null; // if it's an empty string, make it null
+			API_LOG.info("Conversion request received");
+		}
+
+		ConversionReport conversionReport = qrdaService.convertQrda3ToQpp(
+				new InputStreamSupplierSource(originalFilename, inputStream(file), purpose));
+
+		validationService.validateQpp(conversionReport);
+
+		Metadata metadata = audit(conversionReport);
+
+		API_LOG.info("Conversion request succeeded");
+
+		HttpHeaders httpHeaders = new HttpHeaders();
+		httpHeaders.setContentType(MediaType.APPLICATION_JSON_UTF8);
+
+		if (metadata != null) {
+			httpHeaders.add("Location", metadata.getUuid());
+		}
+
+		return new ResponseEntity<>(respond(conversionReport), httpHeaders, HttpStatus.CREATED);
+	}
+
+	protected abstract T respond(ConversionReport report);
+
+	public Metadata audit(ConversionReport conversionReport) {
+		try {
+			CompletableFuture<Metadata> metadata = auditService.success(conversionReport);
+			return metadata == null ? null : metadata.get();
+		} catch (InterruptedException | ExecutionException exception) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either re-interrupt this method or rethrow the "InterruptedException". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)